image-rs/src/layer_store.rs

```diff
@@ -0,0 +1,79 @@
+use std::fs;
+use std::path::PathBuf;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+/// The LayerStore abstracts the image layers storage on the file system.
+pub struct LayerStore {
+    /// OCI image layer data store dir.
+    pub data_dir: PathBuf,
+
+    /// Next layer index
+    layers_index: AtomicUsize,
+}
+
+impl LayerStore {
+    /// Retrieves the next layer index by computing the maximum index from the
+    /// existing layers in data dir.
+    pub fn get_layer_index(data_dir: PathBuf) -> AtomicUsize {
+        let mut next = 0;
+        if data_dir.exists() {
+            if let Ok(paths) = fs::read_dir(data_dir) {
+                for entry_path in paths.filter_map(|result| result.ok()) {
+                    if let Some(name) = entry_path.file_name().to_str() {
```

Comment:
what if the path can't be converted to str? Should we handle the None case?

Reply:
The file name should be an integer which should not fail to convert since it is a valid Unicode value.  If for some reason, the file names cannot be converted, it will not conflict with real layer directories.

Reply:
Added error handling as suggest by @mkulke